reco_utils/recommender/surprise/surprise_utils.py

```diff
@@ -74,7 +74,7 @@ def compute_rating_predictions(
     return predictions.drop(["details", "r_ui"], axis="columns")
 
 
-def compute_ranking_predictions(
+def recommend_k_items(
```

Comment:
This function does not recommend k items. It computes the predictions for all users and items.

Reply:
ohh good catch, the cut of k is done in fact when the metric is computed https://github.com/microsoft/recommenders/blob/master/notebooks/02_model/surprise_svd_deep_dive.ipynb

Reply:
also check with @yueguoguo whether or not we are using the threshold cut https://github.com/microsoft/recommenders/issues/1041

Reply:
if we don't want to work with the scores as a matrix like sar does we could use a heap

```
from heapq import heappush, heappushpop

users = data[usercol].unique()
items = data[itemcol].unique()
preds_lst = np.zeros([len(users), k])

for user_idx, user in enumerate(users):
    heap = []
    for item in items:
        score = algo.predict(user, item).est
        if len(heap) < k:
            heappush(heap, score)
        elif score > heap[k - 1]:
            heappushpop(heap, score);
    preds_lst[user_idx] = heap[::-1]
```

Reply:
`heapq` also has an `nlargest()` method. Not sure which way would be faster (depends on which algorithm it uses).

Reply:
true, but it wasn't clear to me if that method tries to sort the items (which would be unnecessary) or leverages the heap?

Reply:
https://github.com/python/cpython/blob/61b3484cdf27ceca1c1069a351487d2db4b2b48c/Lib/heapq.py#L395
It looks similar to your for loop.